plugin/cog/cog-rangereader-azure/src/main/java/it/geosolutions/imageioimpl/plugins/cog/AzureRangeReader.java

```diff
@@ -0,0 +1,190 @@
+    public byte[] readHeader() {
+        BlobRange blobRange = new BlobRange(0l, (long)headerLength);
+        Response<Flux<ByteBuffer>> response = client
+                .downloadWithResponse(blobRange, null, null, false)
+                .block();
+
+        buffer.position(0);
+        response.value()
+                .map(bb -> {
+                    buffer.put(bb);
+                    return bb;
+                })
+                .blockLast();
+
+        buffer.rewind();
+        byte[] b = new byte[headerLength];
+        buffer.get(b, 0, headerLength);
+        return b;
+    }
+
+
+    @Override
+    public void readAsync(Collection<long[]> ranges) {
+        readAsync(ranges.toArray(new long[][]{}));
+    }
+
+    @Override
+    public void readAsync(long[]... ranges) {
+        ranges = reconcileRanges(ranges);
+
+        Instant start = Instant.now();
+        Map<Long, CompletableFuture<Response<Flux<ByteBuffer>>>> futures = new HashMap<>(ranges.length);
+
+        for (int i = 0; i < ranges.length; i++) {
+            BlobRange blobRange = new BlobRange(ranges[i][0], ranges[i][1]);
+
+            CompletableFuture<Response<Flux<ByteBuffer>>> future = client
+                    .downloadWithResponse(blobRange, null, null, false)
+                    .toFuture();
+                    futures.put(ranges[i][0], future);
+        }
+
+        awaitCompletion(futures);
+        Instant end = Instant.now();
+        LOGGER.fine("Time to read all ranges: " + Duration.between(start, end));
+    }
+
+    protected void writeValue(int position, byte[] bytes) {
+        buffer.position(position);
+        try {
+            buffer.put(bytes);
+        } catch (Exception e) {
+            LOGGER.severe("Error writing bytes to ByteBuffer for source " + uri);
+        }
+    }
+
+    /**
+     * Blocks until all ranges have been read and written to the ByteBuffer
+     *
+     * @param futures
+     */
+    protected void awaitCompletion(Map<Long, CompletableFuture<Response<Flux<ByteBuffer>>>> futures) {
+        boolean stillWaiting = true;
+        List<Long> completed = new ArrayList<>(futures.size());
+        while (stillWaiting) {
+            boolean allDone = true;
+            for (Map.Entry<Long, CompletableFuture<Response<Flux<ByteBuffer>>>> entry : futures.entrySet()) {
+                long key = entry.getKey();
+                CompletableFuture<Response<Flux<ByteBuffer>>> future = entry.getValue();
+                if (future.isDone()) {
+                    if (!completed.contains(key)) {
+                        try {
+                            buffer.position((int)key);
+                            future.get().value().map(bb -> {
+                                buffer.put(bb);
+                                return bb;
+                            }).blockLast();
+
+                            completed.add(key);
+                        } catch (Exception e) {
+                            e.printStackTrace();
```

Comment:
If this fails, 1 or more tiles may become corrupt.  This could lead to an error decoding that tile(s).  Because TIFFImageReader reads 1 tile at a time, I wasn't 100% certain if it would fail on a single failed tile, or if only that tile would not render.  Being optimistic, I didn't fail hard -- I wanted to give the other ranges/tiles a chance to be read and cached in the event they still might be rendered.